Corrections to audio buffer size calculations

marcelofg55 opened the pull request:
Quick explanation: The number of channels wasn't being used in the total buffer_size. A second of sound has actually `sample rate * channels`, so our buffer_size must be equal to `buffer_frames * channels`, ref: https://stackoverflow.com/questions/11048825/audio-sample-frequency-rely-on-channels


Long explanation: I found about this issue while working on the surround fixes.
The calculation for the buffer_size before this PR was:
```
buffer_size = closest_power_of_2(latency * mix_rate / 1000);
samples_in.resize(buffer_size);
buffer_frames = buffer_size / channels;
```
And the new one:
```
buffer_frames = closest_power_of_2(latency * mix_rate / 1000);
buffer_size = buffer_frames * channels;
samples_in.resize(buffer_size);
```
With for example latency=25, mix_rate=44100 and channels=6
Old one:
```
buffer_size = 25 * 44100 / 1000 = 1102 => closest_power_of_2(1102) = 1024
buffer_frames = 1024 / 6 = 170
```

New one:
```
buffer_frames = 25 * 44100 / 1000 = 1102 => closest_power_of_2(1102) = 1024
buffer_size = 1024 * 6 = 6144
```

So the old one gave a buffer size really small with 6 channels (it was also incorrect for 2 or 8 channels of course).

Note: I've modified only alsa, pulseaudio, osx, rtaudio and wasapi drivers, the rest needs to be checked/tested.

marcelofg55 committed:
Corrections to audio buffer size calculations

reduz commented:
This makes sense but, as far as I remember, its very API dependent. Some APIs like buffer size in frames, others in samples, others in bytes, others pay attention to channels and others do not. Did you check every API to make sure this was the case?

akien-mga commented:
We should probably get this tested on as many drivers as possible, and see how it's doing. There would be:
- [ ] Android, JAndroid (not changed yet, also unused)
- [ ] Android, OpenSL (not changed yet)
- [ ] Haiku, MediaKit (not changed yet)
- [ ] iPhone (not changed yet)
- [ ] JavaScript (not changed yet)
- [ ] Linux, Alsa
- [ ] Linux, Pulseaudio
- [ ] Windows, RtAudio
- [ ] Windows, WASAPI
- [ ] OSX
- [ ] UWP, XAudio2 (not changed yet)

marcelofg55 commented:
@reduz Yes that's true, I've check it while coding this and alse re-checked it now just in case:
- With OSX and WASAPI we write directly into the buffers.
- RtAudio uses frames in the openStream function, from its header: bufferFrames A pointer to a value indicating the desired internal buffer size in sample frames.
- ALSA uses frames, with snd_pcm_hw_params_set_period_size_near its val param is in frames (https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m___h_w___params.html#ga9162045265f283c532634506456cab09).
- PulseAudio uses bytes to determine the buffer length in pa_buffer_attr::tlength.

reduz commented:
Alright, should we try to get this tested more, or we should just merge and let others test?

marcelofg55 commented:
@akien-mga Yeah, testing would this would be great, I'll check soon the other drivers that I haven't checked yet.

marcelofg55 commented:
@reduz If we can test the PR a bit more first I guess it would be better, and then merge it to let others test.

vnen commented:
XAudio2 driver should be rechecked too. I won't be surprised if I made some mistakes there.

marcelofg55 commented:
Update: I've continued testing/improving the PR and:
- I've added a DEBUG_ENABLED only code on AudioServer::_driver_process that prints (every 10secs) an average of the audio driver latency between the _driver_process calls, it's also only used if godot gets called with --verbose.
- I've realized that the WASAPI driver uses WASAPI Shared Mode so we can't set the buffer size, this means we don't have any control on the output_latency.
- OSX audio driver wasn't actually setting the buffer size on the audio_unit property, I've fixed this and output_latency now affects the buffer size.

marcelofg55 commented:
@vnen I've just checked the XAudio2 driver and its buffer size calc is correct 👍 . The only thing that I've seen is that the mix_rate is hardcoded to 48000, is there a reason for that?

vnen commented:
@marcelofg55 likely I hardcoded for testing and forgot it like that.

marcelofg55 commented:
@akien-mga I've just re-tested OSX, Windows RtAudio/WASAPI and Linux ALSA/PulseAudio drivers and they all worked fine. Also I've checked the Android/XAudio2 code and the buffer size calc is right in those two. And the last one, iPhone buffer size is hardcoded, but in my opinion we could actually merge OSX/iPhone audio drivers into one, both use the same CoreAudio API, but I'll better leave this for another PR.

akien-mga commented:
Awesome, let's merge and give it broader testing.
